File: packages/voice-sidecar/src/voice_sidecar/sessions.py
"""In-memory voice sessions — bind a browser stream to an opencode session."""

from __future__ import annotations

import secrets
import time
from dataclasses import dataclass


@dataclass
class VoiceSession:
    id: str
    opencode_url: str
    opencode_session_id: str
    directory: str
    agent: str | None
    composer: bool
    terminal_mic: bool
    created_at: float

    def to_dict(self, *, stream_url: str) -> dict:
        return {
            "id": self.id,
            "stream": stream_url,
            "opencode": {
                "url": self.opencode_url,
                "sessionID": self.opencode_session_id,
                "directory": self.directory,
                "agent": self.agent,
            },
            "createdAt": int(self.created_at * 1000),
        }


class VoiceSessionStore:
    def __init__(self) -> None:
        self._sessions: dict[str, VoiceSession] = {}

    def create(
        self,
        *,
        opencode_url: str,
        opencode_session_id: str,
        directory: str,
        agent: str | None,
        composer: bool = False,
        terminal_mic: bool = False,
    ) -> VoiceSession:
        session = VoiceSession(
            id=f"vs_{secrets.token_urlsafe(12)}",
            opencode_url=opencode_url,
            opencode_session_id=opencode_session_id,
            directory=directory,
            agent=agent,
            composer=composer,
            terminal_mic=terminal_mic,
            created_at=time.time(),
        )
        self._sessions[session.id] = session
        return session

    def get(self, voice_id: str) -> VoiceSession | None:
        return self._sessions.get(voice_id)


store = VoiceSessionStore()

File: packages/voice-server/src/sessions.ts
import { randomBytes } from "node:crypto"
import { resolve } from "node:path"

export type VoiceSession = {
  id: string
  opencodeUrl: string
  controlPlaneUrl: string
  opencodeSessionId: string
  directory: string
  agent: string | undefined
  composer: boolean
  createdAt: number
}

export type CreateSessionInput = {
  directory: string
  sessionID?: string
  agent?: string
  composer?: boolean
  server?: string
}

export class VoiceSessionStore {
  #sessions = new Map<string, VoiceSession>()

  create(input: {
    opencodeUrl: string
    controlPlaneUrl: string
    opencodeSessionId: string
    directory: string
    agent: string | undefined
    composer: boolean
  }) {
    const session: VoiceSession = {
      id: `vs_${randomBytes(9).toString("base64url")}`,
      opencodeUrl: input.opencodeUrl,
      controlPlaneUrl: input.controlPlaneUrl.replace(/\/+$/, ""),
      opencodeSessionId: input.opencodeSessionId,
      directory: resolve(input.directory),
      agent: input.agent,
      composer: input.composer,
      createdAt: Date.now(),
    }
    this.#sessions.set(session.id, session)
    return session
  }

  get(voiceId: string) {
    return this.#sessions.get(voiceId)
  }

  drop(voiceId: string) {
    this.#sessions.delete(voiceId)
  }
}

export function streamUrl(base: URL, voiceId: string) {
  const wsBase = base.origin.replace(/^https:/, "wss:").replace(/^http:/, "ws:")
  return `${wsBase}/voice/session/${voiceId}/stream`
}

export function sessionJson(session: VoiceSession) {
  return {
    id: session.id,
    stream: streamUrl(new URL(session.controlPlaneUrl), session.id),
    opencode: {
      url: session.opencodeUrl,
      sessionID: session.opencodeSessionId,
      directory: session.directory,
      agent: session.agent,
    },
    createdAt: session.createdAt,
  }
}

export const voiceSessions = new VoiceSessionStore()
